### PR TITLE
Get attributes from envoy config.

### PR DESCRIPTION
**Pull request by chowchow316**
#84

### CODE REVIEW

File: src/envoy/mixer/http_control.h [hunk 1]
```diff
@@ -37,7 +37,8 @@ typedef std::shared_ptr<HttpRequestData> HttpRequestDataPtr;
 class HttpControl final : public Logger::Loggable<Logger::Id::http> {
  public:
   // The constructor.
-  HttpControl(const std::string& mixer_server);
+  HttpControl(const std::string& mixer_server,
+              const std::map<std::string, std::string>& attributes);
```
Comment (qiwzhang) [inline src/envoy/mixer/http_control.h]: use 

std::map<std::string, std::string>&& attributes

call it by std::move()

to reduce a copy

File: src/envoy/mixer/http_control.cc [hunk 2]
```diff
@@ -134,19 +134,14 @@ void FillRequestInfoAttributes(const AccessLog::RequestInfo& info,
...
 
-  auto source_service = getenv(kEnvNameSourceService.c_str());
-  if (source_service) {
-    source_service_ = source_service;
-  }
-  auto target_service = getenv(kEnvNameTargetService.c_str());
-  if (target_service) {
-    target_service_ = target_service;
-  }
+  source_service_ = attributes.at(kAttrNameTargetService);
```
Comment (qiwzhang) [inline src/envoy/mixer/http_control.cc]: keep the map in the class member.

// the attributes read from the config file.
std::map<std::string, std::string>& config_attributes_;

In the Check() call,  set all of them to the attributes data structure.

File: src/envoy/mixer/http_control.cc [hunk 3]
```diff
@@ -134,27 +134,23 @@ void FillRequestInfoAttributes(const AccessLog::RequestInfo& info,
...
+  config_attributes_ = std::move(attributes);
 }
 
 void HttpControl::FillCheckAttributes(const HeaderMap& header_map,
                                       Attributes* attr) {
   FillRequestHeaderAttributes(header_map, attr);
 
-  SetStringAttribute(kAttrNameSourceService, source_service_, attr);
-  SetStringAttribute(kAttrNameTargetService, target_service_, attr);
+  for (auto attribute : config_attributes_) {
```
Comment (qiwzhang) [inline src/envoy/mixer/http_control.cc]: const auto &

File: src/envoy/mixer/http_control.cc [hunk 4]
```diff
@@ -134,27 +134,23 @@ void FillRequestInfoAttributes(const AccessLog::RequestInfo& info,
 
 }  // namespace
 
-HttpControl::HttpControl(const std::string& mixer_server) {
+HttpControl::HttpControl(
+    const std::string& mixer_server,
+    const std::map<std::string, std::string>&& attributes) {
```
Comment (qiwzhang) [inline src/envoy/mixer/http_control.cc]: remove "const"

File: src/envoy/mixer/http_control.cc [hunk 5]
```diff
@@ -134,27 +134,23 @@ void FillRequestInfoAttributes(const AccessLog::RequestInfo& info,
...
 
-  auto source_service = getenv(kEnvNameSourceService.c_str());
-  if (source_service) {
-    source_service_ = source_service;
-  }
-  auto target_service = getenv(kEnvNameTargetService.c_str());
-  if (target_service) {
-    target_service_ = target_service;
-  }
+  config_attributes_ = std::move(attributes);
```
Comment (qiwzhang) [inline src/envoy/mixer/http_control.cc]: do this in ":" line

File: src/envoy/mixer/http_control.cc [hunk 6]
```diff
@@ -134,27 +134,23 @@ void FillRequestInfoAttributes(const AccessLog::RequestInfo& info,
...
   mixer_client_ = ::istio::mixer_client::CreateMixerClient(options);
 
-  auto source_service = getenv(kEnvNameSourceService.c_str());
-  if (source_service) {
-    source_service_ = source_service;
-  }
-  auto target_service = getenv(kEnvNameTargetService.c_str());
-  if (target_service) {
-    target_service_ = target_service;
-  }
```
Comment (qiwzhang) [inline src/envoy/mixer/http_control.cc]: remove these const string definitions. such as kEnvNameTargetService, and kAttribName

File: src/envoy/mixer/http_filter.cc [hunk 7]
```diff
@@ -96,7 +96,17 @@ class Config : public Logger::Loggable<Logger::Id::http> {
...
     }
 
-    http_control_ = std::make_shared<HttpControl>(mixer_server);
+    std::map<std::string, std::string> attributes;
+    if (config.hasObject("attributes")) {
+      for (const std::string& attr : config.getStringArray("attributes")) {
+        attributes[attr] = config.getString(attr);
+      }
+    } else {
+      log().warn("No attributes is specified in the config: {}", __func__);
```
Comment (qiwzhang) [inline src/envoy/mixer/http_filter.cc]: Not need to log a warn,  just remove the whole else {}.

Comment (kyessenov) [inline src/envoy/mixer/http_filter.cc]: This needs to be extended to include `upstream_cluster`. @rshriram did something similar for the fault injection filter. Basically, the problem is that if you have two upstream clusters "service_a" and "service_b" in the same http connection manager (say via weighted cluster block), you need to send two different sets of attributes about "a" and "b". Maybe make "attributes" a list of JSON objects that have "upstream_cluster" and "attributes" list.

You do not have to do this in the same PR, but send a follow-up PR that allows this customization.

Comment (kyessenov) [inline src/envoy/mixer/http_filter.cc]: I'm also OK with restricting mixer attributes to be same for all service versions. The drawback is that we cannot do gradual config change for the mixer attributes. The whole idea of service versions is that we can customize what set of attributes to send for a subset of the service destination pods.

**Comment by kyessenov** [comment]
@rshriram question about filter priority - we want mixer filter to be the first in the chain since it decides whether to proceed or not based on the Check call. I see that you made fault filter to be first one too. Is it really necessary for the fault filter to be first?

**Comment by qiwzhang** [comment]
@kyessenov This PR has nothing to do with the filter ordering.  We can address your concerns of multiple upstream in separate PR.

File: src/envoy/mixer/http_filter.cc (hunk 7 above)
Comment (qiwzhang) [inline src/envoy/mixer/http_filter.cc]: Open https://github.com/istio/proxy/issues/88 for first issue.  Did not quite understand the second issue yet.  Will open a separate issue once I understand it.

**Pull request merged**